Rework and improve `LS.COM` command

FeralChild64 opened the pull request:
# Description

- Moved `LS.COM` command implementation to `program_ls.cpp` file, cleaned up the code a little.
- Added support for multiple directories / patterns in the command line.
- Command does not display files and directories with Hidden or System attributes anymore, unless `/a` switch is given. 
Rationale: GNU implementation also skips hidden files by default; there are no system files on Unix, and DOS commands which skips hidden files typically skip system files too.


## Related issues

https://github.com/dosbox-staging/dosbox-staging/issues/2956 (partial fix only)


# Manual testing

Played with the command in  various directories, also containing hidden/system files. Sanitizer build did not reveal any issues.


# Checklist

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

MasterO2 commented:
@FeralChild64 Are you familiar with the issue below as well?

https://github.com/dosbox-staging/dosbox-staging/issues/1336

FeralChild64 commented:
@MasterO2 A lot of GNU ls command functionality is missing, I didn't intend to add everything - I primarily want to make our commands handle file attributes properly, support for multiple patterns and directories was actually very easy to add.

kcgen commented on src/dos/program_ls.cpp:
```diff
@@ -22,14 +22,240 @@
+
+	const size_t num_columns = column_widths.size();
+
+	constexpr int ansi_blue  = 34;
+	constexpr int ansi_green = 32;
+
+	auto write_color = [&](const int ansi_color,
+	                       const std::string& txt,
+	                       const uint8_t width) {
+		const auto margin_size = static_cast<int>(width - txt.size());
+		WriteOut("\033[%d;1m%s\033[0m%-*s", ansi_color, txt.c_str(), margin_size, "");
```
> ```c++
> constexpr int ansi_blue  = 34;
> constexpr int ansi_green = 32;
> WriteOut("\033[%d;1m%s\033[0m% ...);
> ```

We can make use of the `convert_ansi_markup()` routine:


```c++
auto write_color = [&](const std::string& colour,
                       const std::string& txt,
                       const uint8_t width) {
	const auto ansi_txt = convert_ansi_markup(colour + txt + "[reset]");
	const auto margin_size = static_cast<int>(width - txt.size());
	WriteOut("%s%-*s", ansi_txt.c_str(), margin_size, "");
};
```

To replace the  to replace the ANSI numbers and escape sequences with the colour tags:

```c++
write_color("[color=light-blue]", name, column_width);
write_color("[color=light-green]", name, column_width);
```

kcgen replied:
> `const auto margin_size = static_cast<int>(width - txt.size());`

Is there any risk of this going negative?

FeralChild64 replied:
Changed the code to use `convert_ansi_markup` - but I'm calling them before the loop starts, to generate `ansi_blue`, `ansi_green` and `ansi_reset` as strings just once.

We shouldn't go negative (or even 0), unless there is a bug in `LS::GetFileNameLengths` or `LS::GetColumnWidths`. Added an assertion there.

kcgen reviewed (commented):
Just a couple suggestions.

It played with some of the patterns and they work selectively:

![2023-10-23_19-34](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/bb203832-52c6-4b20-b3c4-6b4b25d3c21d)

Hopefully the patterns can be more flexible, so people can use it for bulk discovery of files.. like:

`ls */setup.exe`, and find all the games with a setup tool.

FeralChild64 commented on src/dos/program_ls.cpp:
```diff
@@ -22,14 +22,240 @@
+
+	// Sort directory contents
+
+	constexpr bool reverse_order = false;
+	DOS_Sort(dir_contents,
+	         ResultSorting::ByName,
+	         reverse_order,
+	         ResultGrouping::NonFilesFirst);
```
Actually, I intend to rethink this API one day, so that we can support ordering in `DIR` command like in latest Windows: https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/dir

FeralChild64 commented:
> Hopefully the patterns can be more flexible so it can be used for bulk filtering and discovery.. like:
> 
> `ls */setup.exe`, to find all the games with a setup tool.

This would mean quite a serious change to the implementation - and my main goal was to handle Hidden / System attributes in the command.

kcgen commented:
Testing with this directory tree:
```
.
├── EMPTYDIR
├── FILENAME
└── SUBDIR
    └── SUBFILE
```

### No wildcard tests

- [X] `ls` Full name (file)

    ![2023-10-24_11-13](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/98ba951c-31a9-4063-ad54-2dc4b3b2d1f0)

- [x] `ls` Full name (directory)

    ![2023-10-24_11-12](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/ff5bad97-61d5-48a0-a73d-40c3ae748efb)

- [x] `ls` Full name (directory) with slash/

    ![2023-10-24_11-02](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/b7ff323a-6b10-4976-a758-dc836bc9c36b)

- [x] `ls` Full name (directory/file)

    ![2023-10-24_11-16](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/d204535e-8bbd-4c8c-a44c-61d70efd1c07)

### Wildcard tests

- [ ] `ls` prefix wildcard (file)

    Extra matches on DOS side.

    ![2023-10-24_11-18_1](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/816fe959-1fbd-4b8f-8cb3-3a7b8bad7512)

- [x] `ls` postfix wildcard (file)

    ![2023-10-24_11-18](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/1357de16-4f8e-439d-b99c-58f820f2716f)

- [x] `ls` prefix wildcard (directory)

    ![2023-10-24_11-03](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/750e1f58-59d4-49d7-954a-765ea6b88e58)

- [x] `ls` single wildcard (all)

    ![2023-10-24_11-22](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/bbfef091-5682-44b4-b4c0-afdc01d1d258)

- [ ] `ls` single wildcard and slash (all) 

    ![2023-10-24_11-23](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/c62da966-4ee2-4500-8eff-bc21da737302)

- [ ] `ls` wildcard postfix and slash

    ![2023-10-24_11-04](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/64e0cf4d-50c8-49c2-be2b-c76efa8f2312)

- [ ] `ls` wildcard prefix and slash

    ![2023-10-24_11-04_1](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/3b88dba0-cb40-451f-a0e3-57184c14fb4e)


The handling is very inconsistent, to the point where I probably won't know if it's working or not.

> This would mean quite a serious change to the implementation - and my main goal was to handle Hidden / System attributes in the command.

Understood, and this improvement is good.

Just thinking about the unhandled wildcard cases: would it be easy to constraints/block them?

I would be a lot more comfortable using this if it it printed a color warning: **`"Unhandled wildcard pattern"`**, instead of confidently telling me lies, like: `"No files or subdirectories"` :sweat_smile:

FeralChild64 commented:
@kcgen

Latest changes:
- added detection of unsupported wildcards
- changed switch `/a` to `-a` (since I guess you want Unix syntax for this command)

As for `ls *NAME` behavior - this is the result of reusing DOS wildcard support code within `ls` implementation. To fix this, we would need to implement brand new Unix-style wildcard matching.

kcgen commented:
> * changed switch `/a` to `-a` (since I guess you want Unix syntax for this command)

The slash was good (sorry, when I tested: `ls -d` I just typed the command into the DOS side too, without thinking.. and it ignored it).

I don't think we should change the `/arg` syntax on the DOS side.

kcgen commented:
> Latest changes:
> 
>     * added detection of unsupported wildcards
>
> As for `ls *NAME` behavior - this is the result of reusing DOS wildcard support code within `ls` implementation. To fix this, we would need to implement brand new Unix-style wildcard matching.

The new functionality is a lot clearer:

![2023-10-24_13-07](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/40911b0e-b2be-43c7-9fc3-2e894ce1cc70)

I have no doubt about using it now; just read the message and carry on.

If you want to revert/drop the `/a` to `-a` change, I think it's OK to merge.

FeralChild64 commented:
I'll revert the `/a` change, just wait a few minutes.

FeralChild64 commented:
OK, change reverted.
